File: packages/java/src/main/java/com/ailabtools/sdk/response/BaseResponse.java
package com.ailabtools.sdk.response;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Collections;
import java.util.Map;

/** Common metadata included in every API response. */
@JsonIgnoreProperties(ignoreUnknown = true)
public abstract class BaseResponse {
    @JsonProperty("request_id")
    private String requestId;
    @JsonProperty("log_id")
    private String logId;
    @JsonProperty("error_code")
    private Integer errorCode;
    @JsonProperty("error_code_str")
    private String errorCodeString;
    @JsonProperty("error_msg")
    private String errorMessage;
    @JsonProperty("error_detail")
    private ErrorDetail errorDetail;
    private String image;
    private Object ratio;
    @JsonProperty("task_type")
    private String taskType;
    @JsonProperty("task_id")
    private String taskId;
    @JsonProperty("task_status")
    private Integer taskStatus;
    @JsonIgnore
    private Map<String, Object> raw = Collections.emptyMap();

    public String getRequestId() { return requestId; }
    public String getLogId() { return logId; }
    public Integer getErrorCode() { return errorCode; }
    public String getErrorCodeString() { return errorCodeString; }
    public String getErrorMessage() { return errorMessage; }
    public ErrorDetail getErrorDetail() { return errorDetail; }
    public String getImage() { return image; }
    public Object getRatio() { return ratio; }
    public String getTaskType() { return taskType; }
    public String getTaskId() { return taskId; }
    public Integer getTaskStatus() { return taskStatus; }
    public Map<String, Object> getRaw() { return raw; }

    public final void setRaw(Map<String, Object> raw) {
        this.raw = Collections.unmodifiableMap(raw);
    }
}

File: packages/php/src/Response/BaseResponse.php
<?php

declare(strict_types=1);

namespace AILabTools\Response;

abstract class BaseResponse
{
    /** @param array<string, mixed> $raw */
    protected function __construct(
        public readonly array $raw,
        public readonly ?string $requestId = null,
        public readonly ?string $logId = null,
        public readonly ?int $errorCode = null,
        public readonly ?string $errorCodeString = null,
        public readonly ?string $errorMessage = null,
        public readonly ?ErrorDetail $errorDetail = null,
        public readonly ?string $image = null,
        public readonly mixed $ratio = null,
        public readonly ?string $taskType = null,
        public readonly ?string $taskId = null,
        public readonly ?int $taskStatus = null,
    ) {
    }

    /** @param array<string, mixed> $payload
     *  @return array<string, mixed>
     */
    protected static function metadata(array $payload): array
    {
        return [
            'raw' => $payload,
            'requestId' => isset($payload['request_id']) ? (string) $payload['request_id'] : null,
            'logId' => isset($payload['log_id']) ? (string) $payload['log_id'] : null,
            'errorCode' => isset($payload['error_code']) ? (int) $payload['error_code'] : null,
            'errorCodeString' => isset($payload['error_code_str']) ? (string) $payload['error_code_str'] : null,
            'errorMessage' => isset($payload['error_msg']) ? (string) $payload['error_msg'] : null,
            'errorDetail' => isset($payload['error_detail']) && is_array($payload['error_detail'])
                ? ErrorDetail::fromArray($payload['error_detail'])
                : null,
            'image' => isset($payload['image']) ? (string) $payload['image'] : null,
            'ratio' => $payload['ratio'] ?? null,
            'taskType' => isset($payload['task_type']) ? (string) $payload['task_type'] : null,
            'taskId' => isset($payload['task_id']) ? (string) $payload['task_id'] : null,
            'taskStatus' => isset($payload['task_status']) ? (int) $payload['task_status'] : null,
        ];
    }
}
